src/frontend/flight/sync/utils.rs

```diff
@@ -0,0 +1,425 @@
+use crate::frontend::flight::sync::schema::SyncSchema;
+use crate::frontend::flight::sync::writer::DataSyncCollection;
+use arrow::array::{new_null_array, Array, ArrayRef, RecordBatch, UInt64Array};
+use arrow::compute::{concat_batches, take};
+use arrow_row::{Row, RowConverter, Rows, SortField};
+use arrow_schema::SchemaRef;
+use clade::sync::ColumnRole;
+use datafusion::physical_expr::expressions::{MaxAccumulator, MinAccumulator};
+use datafusion_common::{DataFusionError, Result};
+use datafusion_expr::{lit, Accumulator, Expr};
+use itertools::Itertools;
+use std::collections::{HashMap, VecDeque};
+
+// Get rows out of the provided columns, and a null row for comparisons
+fn rows(columns: Vec<ArrayRef>) -> Result<(Rows, Rows)> {
+    let (fields, nulls): (Vec<SortField>, Vec<ArrayRef>) = columns
+        .iter()
+        .map(|array| {
+            let data_type = array.data_type();
+            (
+                SortField::new(data_type.clone()),
+                new_null_array(data_type, 1),
+            )
+        })
+        .collect::<Vec<_>>()
+        .into_iter()
+        .unzip();
+
+    let converter = RowConverter::new(fields)?;
+    Ok((
+        converter.convert_columns(&columns)?,
+        converter.convert_columns(&nulls)?,
+    ))
+}
+
+// Compact a set of record batches into a single one, squashing any chain of changes to a given row
+// into a single row in the output batch.
+// This means that if a row is changed multiple times, only the last change will be reflected in the
+// output batch (meaning the last NewPk and Value role columns and the last Value column where the
+// accompanying Changed field was `true`).
+pub(super) fn compact_batches(
```

Comment:
While I see the point about squashing changes, isn't it pretty wasteful to be converting to rows and then back to columns?

Reply:
Also, can't you squash rows without actually converting to row format?

Reply:
Fair point; for one thing note this was designed to be efficiently encodable: 
https://github.com/apache/arrow-rs/blob/8752e01be642bce205984e16b44e06078413dc68/arrow-row/src/lib.rs#L20-L24

In fact DataFusion itself relies on it for some core computations itself (also we only ever transpose the PKs, though that isn't a big factor anyway).

Reply:
> Or I'm not sure what you mean

Ah, so what needs to happen above is the squashing (compaction) of all chains of changes to a given row (that may encompass multiple input rows), into a single output row, which will then be applied to the target table.

These chains can be of arbitrary lengths, from 1 (meaning a row is touched only once) to the batch row count (meaning the entire batch conveys changes for a single original row). In addition, various things can happen in each chain, such as value changes, pk changes and entire row deletion.

In principle I think this kind of computation can be formalized using recursive CTEs. However there are 3 things to note here:

1. I briefly experimented with recursive CTEs, managing to get some basic (but not full) functionality that we need here. The simplest prototype worked ok, however slightly more involved stuff made DF hang and spike the CPU, meaning there are some bugs there and I can't resolve those now.
2. Even if there were no issues in the execution, the recursive CTE implementation (using dataframe/plan/expression primitives) would be much more complex, since it would approach the problem in a round-about way.
3. Ultimately, even such an implementation would entail using the `arrow-row` under the hood somewhere (e.g. in the grouping/window aggregation phases).

Reply:
Yeah, I think normally we could do this with something like this:

```
WITH latest_rows AS (
  SELECT row_number() OVER (PARTITION BY pk ORDER BY timestamp DESC) AS _row_num
  FROM table
) SELECT [cols] FROM latest_rows WHERE _row_num = 1 AND NOT is_deleted
```

which would still process things row-by-row in the background but at least would reduce the amount of our custom code that we have to write.

But with PK-changing UPDATEs we have to be clever. It does raise the question of why we expect the API user to convert row-based changes into Arrow and then this code transposes it back into rows, then back again into columnar (maybe V2 of this would be Avro changes instead).

We can always see how this performs under benchmarks (append-only, update-intensive, update-intensive with PK changes). Ultimately, we'll have to do this transpose/compact somewhere - on write or on every read.